gr-soapy/grc/soapy_sdrplay_source.block.yml

```diff
@@ -86,17 +106,31 @@ templates:
       self.${id}.set_sample_rate(0, ${samp_rate})
       self.${id}.set_bandwidth(0, ${bandwidth})
       self.${id}.set_antenna(0, ${antenna})
-      self.${id}.set_gain_mode(0, ${agc})
       self.${id}.set_frequency(0, ${center_freq})
       self.${id}.set_frequency_correction(0, ${freq_correction})
-      self.${id}.set_gain(0, min(max(-${gain}, 0.0), 47.0))
+      self.${id}.set_gain_mode(0, ${agc})
+      ##
+      ## Bounds check. Valid values for LNA state (RFGR here)
+      ## vary by model, so runtime changes may cause errors.
+      self.${id}.set_gain(0, 'IFGR', min(max(59 - ${gain}, 20), 59))
+      self.${id}.set_gain(0, 'RFGR', min(max(${lna_state}, 0), 9))
+      ##
+      ## Intercept the agc callback and restore the gain when agc is
+      ## disable. The driver does not do this. LNA state is preserved.
+      def _set_gain_mode(channel, agc):
+          self.${id}.set_gain_mode(channel, agc)
+          if not agc:
+                self.${id}.set_gain(0, 'IFGR', min(max(59 - self.gain, 20), 59))
+      self.set_gain_mode = _set_gain_mode
+
   callbacks:
     - set_sample_rate(0, ${samp_rate})
     - set_bandwidth(0, ${bandwidth})
     - set_antenna(0, ${antenna})
-    - set_gain_mode(0, ${agc})
     - set_frequency(0, ${center_freq})
     - set_frequency_correction(0, ${freq_correction})
-    - set_gain(0, min(max(-${gain}, 0.0), 47.0))
+    - self.set_gain_mode(0, ${agc})
+    - set_gain(0, 'IFGR', min(max(59 - ${gain}, 20), 59))
```

Comment:
I don't know how these callbacks work but it'd be nice to _not_ make this call when `agc=True` since we know it'll just spam the console, and when agc gets unset the value should be reapplied.

Reply:
No hello, spam warnings suppressed.